messaging-spec.md

```diff
@@ -0,0 +1,351 @@
+
+This specification focuses on core messaging functionality and does not include high-level abstractions, prebuilt integrations, or advanced reliability mechanisms beyond the guarantees provided by XCM and ISMP.
+
+## Non-functional Deliverables
+The "API" stands for the api the smart contracts have access to.
+
+- **Stable:** contracts are generally immutable and therefore need a stable interface, especially as the runtime may be evolving over time.
+- **Future Proof:** technologies improve, so the API implementation should be adaptable to new approaches without affecting the interface. The API implementation should also select the optimal approach to realise a use case from a range of possible technical implementations.
+- **Versioned:** not every future technology can be retro-fitted, so the API will be versioned.
+- **Simple:** we want to make the use of the API by smart contract developers as easy to use as possible. A one-to-one mapping to pallet functionality would echo complexity up the stack. Pop API strives to shield dApp developers from this complexity.
+- **Storage Efficient**: The contracts size impacts the cost to deploy.
+- **Test Coverage**: Unit test coverage for each extrinsic at least 80%
+
+## Functional Deliverables
+- A contract can send a message to the pop-net runtime.
+- A contract on pop-net can send an ISMP GET message to a foreign chain.
+- A contract on pop-net can send an ISMP POST message to a foreign chain.
+- A contract on pop-net can send an arbitrary XCM-based message to a foreign chain.
+- A contract on pop-net can register a callback for execution on message response, regardless of message type.
+- A contract on pop-net can query a response regardless of message type.
+- A contract is able to remove request data if they have permission.
+- A contract using Ismp is forced to use signed requests.
+
+There are 3 main areas to consider.
+1. Runtime: Responsible for recording message state and sending messages, index versioning (errors, runtime calls) and initiating callbacks.
+2. Chain extension: Acting as the communication layer between the contract and the runtime. Responsible for dispatching runtime calls and since it has access to the contract environment it is also responsible for charging weight appropriately.
+3. Contract api: A collection of ink abstractions a contract developer can use to easily access exposed runtime features.
+
+## Non-deliverables
+- A high-level abstraction layer for developers (this version focuses on raw ISMP/XCM capabilities)
+- Prebuilt integrations with specific parachains.
+- Automated retries or reliability mechanisms beyond ISMP/XCM’s existing guarantees.
+- Supporting ISMP data processing for RLP encoded contracts.
+- Continuously running relayer (e.g. in pop-node that automatically relays messages).
+- Adding other protocols beyond ISMP and XCM.
+- Drink! e2e testing.
+- Refunding unused weight to developers.
```

Comment:
Depending on how you tackle the weight, this might be kinda easy to do. With fresh eyes, this seems like we should try to make this possible.

Reply:
i have added it to deliverables, i agree with the weights resolved properly this should be ok.